modules/core/src/types/layer-props.ts

```diff
@@ -56,7 +64,7 @@ export type LayerProps<DataType = any> = {
   /**
    * Custom implementation to fetch and parse content from URLs.
    */
-  fetch?: (
+  fetch: (
```

Comment:
Layers can be constructed with multiple props arguments, so all fields will be optional regardless of whether we require it in the types.

Reply:
I struggled with this when migrating carto & quad-key-layer (both in progress) and i concur with this change.
We should have clean indication what is _always_ defined in runtime, otherwise, we have to `assert`  everything in class internals.
So, if `defaultProps` are providing certain property, then typings shouldn't be marked as optional.

(`this.state` is great offender also, lot's of code _just_ assumes it's initialized)

Reply:
I fully agree that we don't want the prop typings to be optional inside the layer implementation, since that requires us to add a lot of null checks even though the props always have a default value.

I am just saying that this can be achieved with the typescript `Required<>` construct. 

```
export type LayerProps = {
  id?: string
};

const DEFAULT_PROPS: Required<LayerProps> = {
  ...
};

export class Layer {
  props: Required<LayerProps>;
  id: string;

  constructor(props: LayerProps) {
    this.props = {...DEFAULT_PROPS, ...props};
     this.id = props.id; // both are unconditional
  }
}
```

The advantage is that this allows us to define types that describe the external API rather than optimize for implementation.

Reply:
After few thoughts i agree, and it would work great, i've even `[played with this idea ](https://github.com/zbigg/deck.gl/commit/a9883a1be3cb320b470564fe6fe5ff98cfa20307)`a bit, but what i see that simply coercing `Required<PropsT>` into `Component<PropsT>.props` just doesn't work. Not sure why

For some reason, in generic classes ... "this pattern" it just [doesn't work](https://www.typescriptlang.org/play?ts=4.0.5#code/FAFwngDgpgBAwgewLYQQOymkAFATgiAZxgF4YBvYGGNAQySgH4AuGQkXASzQHMBuGAHpBMAiE7paAGwA0MAEYBXEDAAWtYgBMoAM1qKpIKjFxRpUsAHkI4yVNbsuvGAB8YitNp3comvsABfYGCAY3R2GC99QzwCQlYAJSgAR0VOU00AHkQUdEwcfCIAPlIKYzoGVgAiAA8aqpljU3MrGwk6e3dPXR9NQODhUUVcGDDcjCwAWilOAGsoYBCpDWIc1AmQTOwYKBqQTE1V5HX82KJStbysM8ISymoIQviYJNT03y2i4OowtEdFEIgBC4AAUjzirGwtFw4mknwAlGVqNRBgAVdQqECqThaBBQQhoADkKgA7sDZjBaJ4YJwQITiAgKR5tLh2FTNLR5FIoHJafSYKhCIROFzYFjaKTYIRoCFONJOAAvWAk2mqURYqC4YwokTg87smAatAwAAGUQMBTiJpgSFoYAUsDQCBUmAQih4aqBOz+w1gVXMAqelNMhtUmqgDVDEv5JNU9pJsHUADdYAxcDxfDcYNqhCIE6MqTAZimYCqsQWImHTDmsTiAHR64hkch11vmmJPOSthtBgICQYaF4pNIZT7fZG-CJpjOaAAiumilqIiWH7yy2BKzYClOIVPtgwzIAAspqZ-O9BabiDa4Q65OOACgbh4ePqDee3FSuQYN3p75z4uNxdq2jbUNugxIMCsAzPMQwjIohCwCEGgLGB446AgCAgoi9zIoMSiYtixA4nBMAAPq2twZHuIhFawAafLEOwnBSFIXReL0NLGuA0CEAAhDmpggMM3FER+RB1hUUB1tyvBYsYQRBIsyxCvAxxXCAAByCAAOKYJqnAhEigYQkObyjpcGw3F8xj3rgj7AmCTyQtCsJSNk6lWU8RQ4TWYmgc23btkuhDAeJxBBGhxgYVhvnIiYUDCbgon1o2kn0NJsk8PJUXUPI0LYcZ1BCSJoapU8dbNKxrS2B0CmBEAA) (link to TS playground), not sure why

Reply:
FWIW, I successfully used the `Required<>` approach consistently across the entire framework for the luma.gl v9 API: https://github.com/visgl/luma.gl/blob/master/modules/api/src/adapter/resources/render-pass.ts#L16

Reply:
I think I've nailed  and that's what i've found:
* this approach doesn't work in any generic where props type contains optional fields like`class Layer<PropsT = extends LayerProps> extends Component<PropsT>` - our case
* this approach works with not-generic classes (like in linked @luma.gl example)

It is caused by `PropsT extends LayerProps` being more permissive that we would like.
In our case, `LayerProps` contains, `{fetch?: something}` and when we say `extends LayerProps` Typescript will allow any type that narrows this, including , `{fetch: undefined}`!.
That means that `Required<{fetch: undefined}> would yield `{fetch: undefined} and thus in all possible instantiations that match `PropsT extends LayerProps` there is one where `fetch` can be indeed `undefined`.

Demonstrated in TS Playground [here](https://www.typescriptlang.org/play?ts=4.6.2#code/PTAECgDsEMFsFMDOAHaBjeoDCBXATnvJAC6gDe4ooxAnspgDLQ3x4AKeA9soqALzlKVUADN4xNAAsA-AC5QACgCU-AHygAbpwCWAEwDcQgL7hTVNJ0iJSu+COg4ANsQ7dE8gErwAjjm2FdAB4mFnYuHnUBCmFRcSl5ZTVyEyoTITRHaEReENZA1x4AFVB4AA9iIl0c5lYC3gFcsLd1aOFkcPdQL19-eCC6wtUhc0trPBw0Yk48BXa3eTZoPGJtaEd8jsGVVpjiSW1EADo5nn5yQ4vbeycXDoAaUAvjjtAjUCyunz8AjbdB4deZmEIk4nESO2EFisnEc8EOjk4AHMFHsDs83IcxBJJMoVCBsNBIJBOKRtJAtABrTCE0CcABGACt4JNQAB3fZSUAHUDITjZbR0xw0UAAchwkCuZL6IoBKUBQnx7Jo0gVYDpzIciEwAH10BY8LoyYjtdROKBCnQkGg8NpkKQtbBCSs0LxOCJQOLJZA+sBuCtLGtgGUKhLeGJoMR8EgucQRbxefzBZgplyrMQnasKhAqLR6F0kTh4I06mdiy8AGSCGJY+IeiV2KW6UD41Gu90WfBa0C6ThISAi0iOqmgLVWeAPOk4UhPAF0pZyEfEG2QRGGVLpTLZfOIwuNErlSrVUKBDwFos1JoRKvCOczbYAkbQ2EI5Gt9E8TFxHFKPFgaCI6AyQeB9QChRAYThF8UX2I5b3hIhET2JRZWMcA0kVTA0BpNZYQ0TNkxg0A6RoB4iEQfAjWoSQI1pfBQAGUBvQ0VhQGo3htRQCjOBwRATQAA09BtvV0PjeHAfELFgSiRC4WBaTtbQA0cURtHgRwqigOAkFQDB6PCPk1gARmvXNGAvOoAEkSFYGBlKiAEa0kBIVD4dQtD0NdhEIHCaAAeQUpSAFkIxtUp5EgHBYHVPAAG0AF1QAAHzrL0+lQoQyl5ZZqEtUAyzcM5FmWVZ1nyngrIqPBbKGIEwJsOwHGcOpPC+XogjKxAKpstZImvKhHOcpIyCMO4AW8xwhX8-1bOCpdtDClKhLS9cNyyI88gY4NDzy8yOi6qq1lLXa3H26q+p5DoWp6H4BiGGI6vGSZplmS76KWFY1l+IpVHvGIcxg996nOS4GpuOoHieE5eDeD5um+PovsQf4YjSatQXBECwIg+EkWgtEoc-bFcWbMAsEJYlSXJThhxpekmRZdltE5bkE0QAUhVFQSREbGU-tA0Zsagt8CfGyaApmkL5vglckJJojLB4h5WUwajmIu7gWNM2l3VF4U-UU2z1eQEDtHdPGjhF+AfKmg21lm0Lfr5-mn0g3HhY6Q5dZtoLJdKaXEMkZC+bldcVqofE6QVxAHm0WNeG0WBWfZ5MzVS0AACJTx3eB09AzdeGI940DQPQiGIWQhC1rPd2O04Glr3hKwhAbFu54S5dbbX+c7TAez7AdQCHTBRy1Ccp0eC5Z3neQxiNTy5QyNbtxr0J9xDKoduPavz1COoWmzYRUKMIA).

So, my proposal _unforunately_ would be to:
 * define all props in type _without_ optionals, naming `SomethingParams`
 * true optionals would be `field: type | undefined | null`
 * clients would be "handed" with exported, `export SomethingProps = Partial<SomethingParams>`
 * internals would work on `SomethingParams` where everything is "mandatory" or "truly optional"

Reply:
I made the same observation when mixing `Required` and generics. I am going to merge the `LayerProps` definition as is (very similar to  @zbigg's proposal) and we can continue to experiment as we type more layers.